guides/hack/51-experimental-features/02-read-only.md

```diff
@@ -0,0 +1,213 @@
+
+## Enabling readonly
+In your relevant file(s), add `<<file:__EnableUnstableFeatures('readonly')>>`.
+
+## How does it work?
+When an Object has the `readonly` keyword applied (e.g. `private readonly Foo $x;`, there are two new constraints on that Object.
+* **Readonlyness:** Object properties can not be modified (i.e. mutated).
+* **Deepness:** All nested properties are readonly.
+
+### Readonlyness
+Object properties can not be modified (i.e. mutated).
+
+``` Hack
+function test(readonly Foo $x) : void {
+  $x->prop = 4; // error, $x is readonly, its properties can not be modified
+}
+```
+
+### Deepness
+All nested properties are readonly.
+
+``` Hack
+class Bar {
+  public function __construct(
+    public Foo $foo,
+  )
+}
+
+class Foo {
+  public function __construct(
+    public int $prop,
+  )
+}
+
+function test(readonly Bar $x) : void {
+  $foo = $x->foo;
+  $foo->prop = 3; // error, $foo is readonly
+}
+```
+
+## Applications
+This is a list of all of the valid locations of the keyword.
+
+### Parameters and return values
+Parameters and return values of any callable can be marked `readonly`.
+
+``` Hack
+function foo(readonly Foo $x): readonly Bar {
+  return $x->bar;
+}
+```
+
+A readonly *parameter* signals that the function/method will not modify that parameter; a readonly *return type* signals that the function returns a readonly reference to an object that can not be modified.
+
+### Static and regular properties
+Static and regular properties can be marked `readonly` and, when applied, can not be modified.
+
+``` Hack
+class Foo {
+  private readonly Bar $bar;
+  private static readonly Bar $static_bar;
+}
+```
+
+#### Other restrictions
+Accessing a readonly property (i.e. a property that’s marked readonly at the declaration, not accessing a property off of a readonly object) requires readonly annotation.
+
+``` Hack
+class Foo {
+  public function __construct(
+    readonly Bar $bar,
+  )
+}
+
+function test(Foo $f): void {
+  $bar = readonly $f->bar; // this is required
+}
+```
+
+And, class properties can not be set to readonly values unless they are readonly properties.
+
+``` Hack
+class Foo {
+  readonly Bar $ro_prop;
+  Bar $mut_prop;
+}
+
+function test(
+  Foo $x,
+  readonly Bar $bar,
+) : void {
+  $x->mut_prop = $bar; // error, $bar is readonly but the prop is mutable
+  $x->ro_prop = $bar; // ok
+}
+```
+
+### Lambdas and function type signatures
+`readonly` is allowed on inner parameters and return types on function typehints.
+
+``` Hack
+function call(
+    (function(readonly Bar) : readonly Bar) $f,
+    readonly Bar $arg,
+   ) : readonly Bar {
+   return $f($arg);
+}
+```
+
+### Expressions
+`readonly` can appear on expressions.
+
+``` Hack
+function foo(): void {
+  $x = new Foo();
+  $y = readonly $x;
+}
+```
+
+### Functions / Methods
+`readonly` can appear as a modifier on instance methods.
+
+``` Hack
+class C {
+  public readonly function bar(): Bar {
```

Comment:
I guess this is more about feature design than docs: any note we can add on 'readonly' vs pure or other coeffects here?

Reply:
Added a note on coeffects affecting the full function scope v. readonly only affecting expressions / values.